test-#5: Improved code for api handlers and added unit tests for backend controllers using jest

hemant14050 opened the pull request:
## Summary
This PR is a continuation of PR #48 

## Description
In this PR,
1. Improved project structure by moving controllers from the ` routes ` folder to a dedicated ` controllers ` folder. The existing controller functions have been moved from the routes folder to this new location. 
2. Created setup for tests adding unit using jest and setup for babel, prettier
3. Added unit tests for various backend controller functions (as discussed in PR #48)
4. Improved code by adding validations in the required controllers (as discussed in PR #48)
       - validation for "All fields are required"
       - validation for "image url extensions"
       - validation for "select categories not more that 3"
       - while search by category - ensure category is valid
6. Improved controllers to pass all required testcases

## Images
![image](https://github.com/krishnaacharyaa/wanderlust/assets/112842857/f2933136-19b8-4887-8749-51ee867e32c9)
![image](https://github.com/krishnaacharyaa/wanderlust/assets/112842857/d3ff0182-7e7e-4c01-b0f6-fd6c1b181e19)

## Issue(s) Addressed
Closes #5 

## Prerequisites

- [x] Have you followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions) ?

hemant14050 commented:
Hi, @krishnaacharyaa and @chinmaykunkikar,
I've addressed the feedback from the previous review and made changes.
Please take a look and let me know if there are any further suggestions or changes needed. Thank you for your feedback!

krishnaacharyaa commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -0,0 +1,424 @@
+      categories: [validCategories[0]],
+      description: 'This is a test post.',
+      isFeaturedPost: false,
+    };
```
Can we create a private function which returns us the Post which takes in the attributes via params and use it everywhere, right now it is too much spread around the place.

krishnaacharyaa commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -0,0 +1,424 @@
+      description: 'This is a test post.',
+      isFeaturedPost: false,
+    };
+    const req = { body: postObject };
```
We can replace this with private function

krishnaacharyaa commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -0,0 +1,424 @@
+  });
+
+  it('should handle errors - Internal Server Error', async () => {
+    const postObject = {
```
Can we refactor all the "should handle errors" to use a private function ? Please try to explore on that part if there is any refactoring possible, if there is no option we can carry on with this.

krishnaacharyaa commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -0,0 +1,424 @@
+    const req = {};
+
+    const mockFeaturedPosts = [
+      {
```
Use private function in all the relevant places :)

krishnaacharyaa commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -0,0 +1,424 @@
+  });
+
+  it('should handle errors', async () => {
+    const req = { params: { category: 'Travel' } };
```
If we cannot refactor, kindly mention the specification of what type of error needs to be handled as done in the above test cases

krishnaacharyaa commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -0,0 +1,424 @@
+  });
+
+  it('should handle errors', async () => {
+    const req = { params: { id: '6910293383' } };
```
Same here, please look into all the occurrences

krishnaacharyaa commented on backend/utils/constants.js:
Is this files created purely for tests ? if yes kindly place this inside the tests folder, else if it is possible to use this in our addblog page in frontend, replace frontend to use this constants :)

hemant14050 replied:
No, not for tests only, constants file is created for both test and controller of find by category handler to check whether params category is valid or invalid.
I think there is similar file for frontend and now I have created new for backend now :)
To be honest, I didn't got the point what to replace in frontend:)

krishnaacharyaa replied:
Okay, got it, my bad I thought we were missing this in the frontend, now i now we have it even in frontend, okay then we are good in this part, no need to change anything, as we will have isolation between both frontend and backend, sounds good to me :)

krishnaacharyaa reviewed (changes requested):
Hey @hemant14050, I really appreciate, you taking up the task and successfully completing it, Thank you. Just some minor refactoring please look into that, we are almost good to merge :)

hemant14050 commented:
Hi @krishnaacharyaa,
I have refactored code by adding private function to do repetitive work of post object creation and mentioned the specific type of error needs in tests.
Also, added some more tests for error handling. Please review it and let me know if any mistakes or improvement :)

krishnaacharyaa commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -79,32 +85,22 @@ describe('createPostHandler', () => {
   });
 
   it('should handle errors - Internal Server Error', async () => {
-    const postObject = {
```
Can we replace the type of error with the purpose of error, because I see many places it is same "should handle errors- Internal server error", we can't actually differentiate just by reading the title, because all the titles are same, so can we change this to tell us the intention of the test case than just type of exception to be thrown, as already done in the above test cases, like "too many categories", "Missing form fields" this helps us understand the purpose of the tests, so can we change this to "handle errors on post creation failure" and please make this change everywhere we need to change.

hemant14050 replied:
Thank you for your valuable suggestion. I completely agree that having more descriptive test case titles would greatly improve readability and understanding. 
Can I update the test descriptions as below?? 
inside "createPostHandler", replacing "should handle errors - Internal Server Error" with more explicit descriptions like "should handle errors on post creation failure".
Please feel free to give me your suggestion with example on this.

krishnaacharyaa replied:
Thank you @hemant14050 for presenting your ideas, just one suggestion, can we use something like "Post creation: Failure - Invalid image url", basically the template being "{Use Case}:{Success/Failure} - {Description}"

I tried to gather best of worlds for our requirement, so in short we get to know what the test case is all about :)
You can refer: https://medium.com/practical-software-testing/write-better-test-cases-titles-e2989d614b5b
https://www.testmonitor.com/blog/how-to-write-effective-test-case-names-with-examples

hemant14050 replied:
Thank you, @krishnaacharyaa! Your suggestion regarding structured test case titles is truly insightful. After reviewing the blogs, I'm now planning to implement this format soon :)

krishnaacharyaa commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -113,33 +109,21 @@ describe('createPostHandler', () => {
 
 describe('getAllPostsHandler', () => {
   it('should get all posts', async () => {
-    const req = {};
```
Can we try to get consistency in the description like :

get all posts - success 
get all posts -failure 

OR

[POSITIVE] - get all posts 
[NEGATIVE ]- get all posts

Feel free to come up with new version of your self, or something like above we can stick on to, so that when we see in the console, we exactly know what is happening :)

krishnaacharyaa replied:
Thank you @hemant14050, can you once confirm, if you wanna take this up as a refactoring issue and raise new PR ? We can merge this, This already is very useful and serves the very purpose of having the test cases for our application

hemant14050 replied:
Hey @krishnaacharyaa, I'll take up the task of refactoring and complete it soon. So, I think no need for new PR.

krishnaacharyaa replied:
Glad to hear @hemant14050, Thank you for confirming :)

krishnaacharyaa commented on backend/tests/unit/controllers/posts-controller.test.js:
```diff
@@ -245,16 +195,16 @@ describe('getPostByCategoryHandler', () => {
-  it('should handle errors', async () => {
-    const req = { params: { category: 'Travel' } };
+  it('should handle errors - Internal Server Error', async () => {
+    const req = createRequestObject({ params: { category: validCategories[1] } });
```
Please check everywhere, the change is needed.

Pro tip: If it soo happens that you cannot describe the test cases properly even after thinking, then we can delete that test case :)

krishnaacharyaa reviewed (commented):
Thank you for making the changes @hemant14050 , now it looks more clean, just small couple of changes if we could do, we are almost good to merge this PR :)

Let me know if you want to create another PR for refactoring, I'm open for that too, I understand you have put some very decent efforts on this one. Kindly let me know your thoughts, we can do anyway according to your convenience :)

krishnaacharyaa commented:
Hi @hemant14050, can we conclude on this, if you want me to create a new issue for this comments, i'll happy to do so...
Kindly let me know your thoughts

krishnaacharyaa commented on backend/package.json:
```diff
@@ -9,14 +9,23 @@
     "eslint-config-prettier": "^9.0.0",
-    "prettier": "^3.1.0"
+    "jest": "^29.7.0",
+    "prettier": "^3.0.3"
```
Here can you kindly revert back to 3.1.0 it was updated recently the prettier version, might have missed during the merge conflict

krishnaacharyaa replied:
Resolved myself no worries :)

krishnaacharyaa reviewed (approved):
Merging now, thank you for the much needed test cases :)
